examples/frontend-re-frame/src/cljs/frontend_re_frame/core.cljs

```diff
@@ -0,0 +1,157 @@
+(ns frontend-re-frame.core
+  (:require
+   [re-frame.core :as re-frame]
+   [reagent.core :as reagent]
+   [reitit.core :as r]
+   [reitit.coercion :as rc]
+   [reitit.coercion.spec :as rss]
+   [reitit.frontend :as rf]
+   [reitit.frontend.controllers :as rfc]
+   [reitit.frontend.easy :as rfe]))
+
+;;; Events ;;;
+
+(re-frame/reg-event-db
+ ::initialize-db
+ (fn [_ _]
+   {:current-route nil}))
+
+(re-frame/reg-event-fx
+ ::navigate
+ (fn [db [_ route]]
+   ;; See `navigate` effect in routes.cljs
+   {::navigate! route}))
+
+(re-frame/reg-event-db
+ ::navigated
+ (fn [db [_ new-match]]
+   (assoc db :current-route new-match)))
+
+;;; Subscriptions ;;;
+
+(re-frame/reg-sub
+ ::current-route
+ (fn [db]
+   (:current-route db)))
+
+;;; Views ;;;
+
+(defn home-page []
+  [:div
+   [:h1 "This is home page"]
+   [:button
+    ;; Dispatch navigate event that triggers a (side)effect.
+    {:on-click #(re-frame/dispatch [::navigate ::sub-page2])}
+    "Go to sub-page 2"]])
+
+(defn sub-page1 []
+  [:div
+   [:h1 "This is sub-page 1"]])
+
+(defn sub-page2 []
+  [:div
+   [:h1 "This is sub-page 2"]])
+
+;;; Effects ;;;
+
+;; Triggering navigation from events.
+(re-frame/reg-fx
+ ::navigate!
+ (fn [k params query]
+   (rfe/push-state k params query)))
+
+;;; Routes ;;;
+
+(defn href
+  "Return relative url for given route. Url can be used in HTML links."
+  ([k]
+   (href k nil nil))
+  ([k params]
+   (href k params nil))
+  ([k params query]
+   (rfe/href k params query)))
+
+(def routes
+  ["/"
+   [""
+    {:name      ::home
+     :view      home-page
+     :link-text "Home"
+     :controllers
+     [{;; Do whatever initialization needed for home page
+       ;; I.e (re-frame/dispatch [::events/load-something-with-ajax])
+       :start (fn [& params](js/console.log "Entering home page"))
+       ;; Teardown can be done here.
+       :stop  (fn [& params] (js/console.log "Leaving home page"))}]}]
+   ["sub-page1"
+    {:name      ::sub-page1
+     :view      sub-page1
+     :link-text "Sub page 1"
+     :controllers
+     [{:start (fn [& params] (js/console.log "Entering sub-page 1"))
+       :stop  (fn [& params] (js/console.log "Leaving sub-page 1"))}]}]
+   ["sub-page2"
+    {:name      ::sub-page2
+     :view      sub-page2
+     :link-text "Sub-page 2"
+     :controllers
+     [{:start (fn [& params] (js/console.log "Entering sub-page 2"))
+       :stop  (fn [& params] (js/console.log "Leaving sub-page 2"))}]}]])
+
+(defn on-navigate [new-match]
```

Comment:
It is best to not use `subscribe` outside render or other subscriptions, it will create RAtoms which will be only used once etc.
I think it is better to handle this old-match, apply-controllers etc. in `::navigated` event handler, where it  is easy to read the old match from app-db etc.

Reply:
Good point! 👍